modules/client/src/client-common/modules/ipfs.ts

```diff
@@ -3,48 +3,61 @@ import { Context } from "../../client-common/context";
 import { IClientIpfsCore } from "../interfaces/core";
 import { Client as IpfsClient } from "@aragon/sdk-ipfs";
 
-export class IPFSModule implements IClientIpfsCore {
-  private _ipfs: IpfsClient[] = [];
-  private _ipfsIdx: number = -1;
+const IPFSMap = new Map<IPFSModule, IpfsClient[]>();
+const IPFSIdxMap = new Map<IPFSModule, number>();
 
+export class IPFSModule implements IClientIpfsCore {
   constructor(context: Context) {
     if (context.ipfs?.length) {
-      this._ipfs = context.ipfs;
-      this._ipfsIdx = Math.floor(Random.getFloat() * context.ipfs.length);
+      IPFSIdxMap.set(this, Math.floor(Random.getFloat() * context.ipfs.length));
```

Comment:
```suggestion
      // Storing client data in the private module's scope to prevent external mutation
      IPFSIdxMap.set(this, Math.floor(Random.getFloat() * context.ipfs.length));
```